### DISCUSSION TITLE
API design of adding `filter` and `idFilter` in `load` and `transform` hook

### DISCUSSION THREAD

**Discussion by hyfdev** [rolldown/rolldown#1970] [Core Minutes]
```js
import { defineConfig } from 'rolldown'

export default defineConfig({
  input: './index.js',
  plugins: [
    {
      transform: {
        // `filter` actually means filter based on `ModuleType`
        filter: 'jsx',
        // `idFilter` actually means filter based on module id
        idFilter: '*.jsx',
       filter: {
         id: { include, exclude },
         moduleType: { include, exclude },
         code: { include, exclude },
       }
        handler: () => {},
      },
    },
    {
      transform: {
        filter: ['js', 'jsx', /* custom module type */ 'scss'], // Not regex, not glob. Exact match.
        idFilter: ['*.js', '*.jsx', '*.scss', /\.less/], // Support regex
        handler: () => {},
      },
    },
    {
      transform: {
        filter: {
          include: ['js', 'jsx'],
          // `exclude` is kind of useless in `ModuleType`-based filter
          exclude: ['css'],
        },
        idFilter: {
          include: ['*.js', '*.jsx'],
          exclude: ['*.css'],
        },
        handler: () => {},
      },
    },
    // load hooks
    {
      load: {
        filter: ['js', 'jsx'], // Support `filter` based on `ModuleType` too.
        idFilter: '*.js',
        handler: () => {},
      },
    },
    {
      load: {
        idFilter: ['*.js', '*.jsx'],
        handler: () => {},
      },
    },
    {
      load: {
        filter: {},
        idFilter: {
          include: ['*.js', '*.jsx'],
          exclude: ['*.css'],
        },
        handler: () => {},
      },
    },
  ],
})

```

**Comment by hyfdev** [comment]
Using filter based on module id requires adding `id` prefix. This is because we want to encourage users to use `ModuleType`-based `filter` as the initial choice.

**Comment by IWANABETHATGUY** [comment]
should support regex,  glob pattern in `idFilter`

**Reply by hyfdev** [reply]
- Value of `filter` is doing exact match.
 - Value of `idFilter` is doing glob match by default.
 
Regex doesn't affect the overall API form. You could just replace above all `string` to regex.

**Reply by IWANABETHATGUY** [reply]
Yeah, you don't clarify the type before.

**Reply by hyfdev** [reply]
Added it in the content.

**Reply by IWANABETHATGUY** [reply]
:+1:

**Comment by hyfdev** [comment]
`renderChunk` ... hooks will have their filter if needed. But it's not the timings to discuss them here. We need to focus on the important and planned things here.

**Comment by IWANABETHATGUY** [comment]
Also, we might support these two `fields` in `plugin` too e.g.

```js
import { defineConfig } from "rolldown";

export default defineConfig({
	input: "./index.js",
	plugins: [
		{
			// `filter` actually means filter based on `ModuleType`
			filter: "jsx",
			// `idFilter` actually means filter based on module id
			idFilter: "*.jsx",
			transform() {},
			load() {},
		},
	],
});

```
which is equivalent to: 
```js
import { defineConfig } from "rolldown";

export default defineConfig({
	input: "./index.js",
	plugins: [
		{
			transform: {
				// `filter` actually means filter based on `ModuleType`
				filter: "jsx",
				// `idFilter` actually means filter based on module id
				idFilter: "*.jsx",
				handler() {},
			},
			load: {
				// `filter` actually means filter based on `ModuleType`
				filter: "jsx",
				// `idFilter` actually means filter based on module id
				idFilter: "*.jsx",
				handler() {},
			},
		},
	],
});

```

**Reply by IWANABETHATGUY** [reply]
the filter in `hook` should have higher priority, so user could override `filter` for a specific hook

**Reply by hyfdev** [reply]
Other hooks might also have their own filter logic(that's not based on `ModuleType`) and reuse the `filter` name. It's hard to distinguish them in the top level.

**Comment by underfin** [comment]
The name maybe has some confused. How about like this.
``` .ts
import { defineConfig } from "rolldown";

export default defineConfig({
	input: "./index.js",
	plugins: [
		{
			transform: {
				// `moduleType` actually means filter based on `ModuleType`
				moduleType: "jsx",
				// `filter` actually means filter based on module id
				filter: "*.jsx",
				handler() {},
			},
		},
	],
});

```

**Reply by hyfdev** [reply]
This actually looks nice. What I'm worried about is without reading the docs, users' first thought would be use id-based `filter`. Because `filter` is a quite meaningful word, but understanding `moduleType` requires more initiative.

I kind of hope we could force people to use `ModuleType`-based filter in the first place.

**Reply by hyfdev** [reply]
I looked again. Other than can't forcing people to use ModuleType-based filter in the first place, this pattern seems perfect.

**Comment by IWANABETHATGUY** [comment]
IMO, one `filter` field is enough, 
```js
import { defineConfig, moduleTypeFilter } from "rolldown";

export default defineConfig({
	input: "./index.js",
	plugins: [
		{
			transform: {
				filter: [moduleTypeFilter("js"), "*.jsx", /\.ts$/],
				handler() {},
			},
		},
	],
});

```

**Reply by IWANABETHATGUY** [reply]
moduleTypeFilter: 
```js
function moduleTypeFilter(ty) {
  return {type: ty}
}
```

**Comment by hyfdev** [comment]
Having a `exclude` filed on `ModuleType`-based filter seems useless.

**Comment by hyfdev** [comment]
There is also a requirement on filter that based on the content: https://github.com/vitejs/vite/blob/67005949999054ab3cd079890ed220bc359bcf62/packages/vite/src/node/plugins/importMetaGlob.ts#L77

**Reply by hi-ogawa** [reply]
Sharing some examples of the pattern like `code.include(someString)` or `someRegex.test(code)` to bail out transform early:
- Vitest's `vi.mock` transform ([link](https://github.com/vitest-dev/vitest/blob/85fb94a3081558b01f47bd763ccdaeb5df1b98cb/packages/vitest/src/node/hoistMocks.ts#L81-L95)), `vite-env-only` ([link](https://github.com/pcattori/vite-env-only/blob/45a39e7fb52e9fae4300e95aa51ed0880374f507/src/index.ts#L12-L13))
  - the general pattern of these cases is that a plugin has a "runtime" part to be imported in the source codes. For such plugins, they usually have a easier way to detect (at least falsify) "runtime" usage without full ast parsing on its own.
- `"use client"`, `"use server"` directives ([link](https://github.com/hi-ogawa/vite-plugins/blob/34fcdfed68822235fc1a25ae8b11985f8ca8d381/packages/react-server/src/features/client-component/plugin.ts#L144-L154))
  - this also allows early bail out by `code.includes(directiveString)`

**Comment by hi-ogawa** [comment]
I've asked something similar before, but putting it again here for visibility.
What's the ideal filtering for a following use case?

```ts
export default defineConfig({
  plugins: [
    {
      transform: {
        // filter: ??
        // idFilter: ??
        handler: (code, id) => {
          // parse/analyze/transform `code` assuming it's a plain javascript
          // where id could be originally jsx/ts/tsx
          
          // for example, running es-module-lexer 
          esModuleLexer.parse(code); 
        }
      }
    }
  ]
})
```

On Vite, this magically works due to builtin esbuild plugin transforming jsx/ts/tsx before user plugins.

On Rolldown, is it possible to use builtin transformer to change `moduleType` from `jsx/js/tsx` to `js` so that user plugins can chain it and only require module type filter `js` for this use case?

**Reply by IWANABETHATGUY** [reply]
Yes, it could. Rolldown has already overwrite the module-type to `js` when the user transform `jsx/tsx/ts` into `js` using `builtin:transform` plugin

**Reply by hi-ogawa** [reply]
Wow, that's great. I didn't know builtin transform is ready already.

**Reply by IWANABETHATGUY** [reply]
related code https://github.com/rolldown/rolldown/blob/1de875126a80618f24a6e9813dc8cb09b55a5d73/crates/rolldown_plugin_transform/src/lib.rs#L110-L115

**Reply by hyfdev** [reply]
That would be 

```js
export default defineConfig({
  plugins: [
    builtinTransform(),
    {
      transform: {
        filter: ['js'], // Wrting this is a little bit uselss, since after `builtin:transform` all jsx, ts/x become `js`.
        // idFilter: ??
        handler: (code, id) => {
          // parse/analyze/transform `code` assuming it's a plain javascript
          // where id could be originally jsx/ts/tsx
          
          // for example, running es-module-lexer 
          esModuleLexer.parse(code); 
        }
      }
    }
  ]
})
```

**Comment by yyx990803** [comment]
Possible nested API:
```
filter: {
  id:
  moduleType:
  code: 
}
```

**Comment by TheAlexLichter** [comment]
Closing here as [plugin hook filters exist](https://rolldown.rs/guide/plugin-development#plugin-hook-filters) by now.
